### PR TITLE
[BEAM-9421] Website section that describes getting predictions using AI Platform Prediciton

### PR DESCRIPTION
**Pull request by kamilwu**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kamilwu** [comment]
R: @Ardagan

**Comment by Ardagan** [comment]
@aaltay @soyrice

File: website/src/documentation/patterns/ai-platform.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+```
+{"input": "the quick brown"}
+{"input": "la bruja le"}
+``` 
+
+Here is the code:
+
+{:.language-java}
+```java
+// Getting predictions is not available for Java.
```
Comment (Ardagan) [inline website/src/documentation/patterns/ai-platform.md]: We can create tracking Jira for this and add reference.

**Comment by aaltay** [comment]
According to https://github.com/tensorflow/tfx-bsl

> This package is not intended for direct use by TFX users, and its APIs should be considered internal to TFX (therefore there is no backward or forward compatibility guarantee) unless otherwise remarked.

Maybe we need to wait until tfx_bsl is available as a public solution. @yifanmai @katsiapis  -- please advise on when/how we should move forward.

**Comment by kamilwu** [comment]
I see. In that case, I'll be waiting for suggestions on how we should move forward with the documentation.

**Comment by soyrice** [comment]
This looks great. I can give it an editorial pass, but I'll wait until we have a plan to move forward re: tfx_bsl.

In the meantime, here are a couple suggestions for improving the write-up:
- Avoid "we" pronouns wherever possible (same for "let's"). Many of these sentences can be rewritten without pronouns. For instance, "you have to make" can be simplified to just "make."
- If a pronoun is necessary, use the second person "you" to talk directly to the reader.
- Try to keep the opening paragraphs as short and succinct as possible. Consider breaking up long paragraphs into multiple, smaller ones. This makes it easier for readers to scan the doc.
- Introduce the code sample by explaining what it does, rather than explaining afterwards.

I put together a [checklist/template for Beam patterns](https://docs.google.com/document/d/1lawWkFPDYV9ORUZbLZUxTvjG6hunyxDdR4s3tvQi6G4/edit) if you want to take a look. The PR already looks good from a content perspective, so this is just for reference :)

**Comment by kamilwu** [comment]
@soyrice thanks, those are very useful tips! I've pushed new changes, hope it looks even better.

File: website/src/documentation/patterns/ai-platform.md (hunk 1 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Done.

File: website/src/documentation/patterns/ai-platform.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    <li data-type="language-java">Java SDK</li>
+    <li data-type="language-py" class="active">Python SDK</li>
+  </ul>
+</nav>
+
+## Getting predictions
+
+This section shows how to use a cloud-hosted machine learning model to make predictions about new data using Google Cloud AI Platform Prediction within Beam's pipeline.
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library that provides `RunInference` Beam's PTransform. `RunInference` is a PTransform able to perform two types of inference. One of them can use a service endpoint. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for each element, sends a request to Google Cloud AI Platform Prediction service. The transform produces a PCollection of type `PredictLog` which contains predictions.
```
Comment (wenchenglu) [inline website/src/documentation/patterns/ai-platform.md]: "one of them can use a service endpoint", just out of curiosity, what is the other type of inference

File: website/src/documentation/patterns/ai-platform.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+Before getting started, deploy a machine learning model to the cloud. The cloud service manages the infrastructure needed to handle prediction requests in both efficient and scalable way. Only Tensorflow models are supported. For more information, see [Exporting a SavedModel for prediction](https://cloud.google.com/ai-platform/prediction/docs/exporting-savedmodel-for-prediction).
```
Comment (wenchenglu) [inline website/src/documentation/patterns/ai-platform.md]: Does that mean users will need to separately deploy a model first? will it be a better user experience if some initial setup stage for Beam can call AI Platform prediction public API to deploy model and get the service endpoint? 

Also, for batch inference scenario, model deployment is a one-off job, users then need to un-deploy models to avoid unnecessary charges. Should they do that separately, or is there a BEAM final stage we could plug in a API call to delete that model deployment?

File: website/src/documentation/patterns/ai-platform.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  dict_ = json.loads(json_obj)
+  for name, text in dict_.items():
+      value = tf.train.Feature(bytes_list=tf.train.BytesList(
+        value=[text.encode('utf-8')]))
+      feature = {name: value}
+      return tf.train.Example(features=tf.train.Features(feature=feature))
+
+with beam.Pipeline() as p:
+     _ = (p
+         | beam.io.ReadFromText('gs://my-bucket/samples.json')
```
Comment (wenchenglu) [inline website/src/documentation/patterns/ai-platform.md]: is this a single inference input item? or a set of items? FYI, I think AI Platform Prediction supports both. For the latter one, a single HTTP request will embed multiple input items, which might provide better throughput once AI Platform Prediction enables batching mode in their model server.

File: website/src/documentation/patterns/ai-platform.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+         | beam.Map(convert_json_to_tf_example)
+         | RunInference(
```
Comment (wenchenglu) [inline website/src/documentation/patterns/ai-platform.md]: Is RunInference a HTTP call? is there a plan to support gRPC in the future?

File: website/src/documentation/patterns/ai-platform.md (hunk 2 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Offline inference from a SavedModel instance. I decided it's not worth mentioning here, since the article is focused on AI Platform patterns.

File: website/src/documentation/patterns/ai-platform.md (hunk 5 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Yes. The transform uses this client library: https://github.com/googleapis/google-api-python-client. 

> is there a plan to support gRPC in the future?

As far as I know, there is no such plan at the moment

File: website/src/documentation/patterns/ai-platform.md (hunk 4 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: `RunInference` uses a Beam's built-in transform `BatchElements`, which batches elements (it means that it consumes a PCollection of `tf.train.Example` and produces a PCollection of `List[tf.train.Example]`). Thanks to that, multiple input items could be sent using a single HTTP request.

File: website/src/documentation/patterns/ai-platform.md (hunk 3 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Yes, the transform expects that a specified model is up and running. Users should also un-deploy redundant model manually.

It is theoretically possible to implement such initial setup and clean-up operations, however, I wonder whether this is a good use-case for Beam: those are long-running operations and we'd have to wait until they are finished. As a result, Dataflow workers (or any other runner) would be idle for some time.

Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: I'm curious what others think about this. @aaltay @Ardagan

Comment (wenchenglu) [reply website/src/documentation/patterns/ai-platform.md]: I am also curious about the best practice in similar scenarios for Beam customers. When they need to setup something, which can take about 10 minutes, do they normally do that setup manually or as part of pipeline.

For up-deploy, we are considering autoscaling down to 0 when batch job is done, then customers don't need to do that un-deploy manually, neither we need to figure out the Beam clean-up phase.

Comment (aaltay) [reply website/src/documentation/patterns/ai-platform.md]: Beam does not have hooks for job start and job finish. One way to emulate those would be using an airflow wrapper to run some setup and cleanup tasks before and after a task. I do not believe we have a recommendation here but an example pattern might be useful (@rezarokni -- in case this will be a good pattern to add to the backlog.)

For this version, it is probably fine to assume that the life cycle of the model will be managed by the user separately. For a next phase, we can try to build a graph similar to IOs (like BQ IO) to execute a certain setup and clean task in a single worker. (Although we cannot guarantee that it will not be retried)

Comment (wenchenglu) [reply website/src/documentation/patterns/ai-platform.md]: LGTM

**Comment by wenchenglu** [comment]
Thanks for addressing my comments. These changes look good to me. I will ask another engineer from AI Platform prediction to do another pass.

File: website/src/documentation/patterns/ai-platform.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+Once a machine learning model is deployed, prepare a list of instances to get predictions for. 
+
+Here is an example of a pipeline that reads input instances from the file, converts JSON objects to `tf.train.Example` objects and sends data to the service. The content of a file can look like this:
+
+```
+{"input": "the quick brown"}
+{"input": "la bruja le"}
+``` 
+
+The example creates `tf.train.BytesList` instances, thus it expects byte-like strings as input, but other data types, like `tf.train.FloatList` and `tf.train.Int64List`, are also supported by the transform. To send binary data, make sure that the name of an input ends in `_bytes`.
```
Comment (limingxi) [inline website/src/documentation/patterns/ai-platform.md]: I think it would be good to list all possible input formats and output formats here or somewhere else for reference. And for the last sentence, do you mean that we need to change l74 to something like: 
feature={name+'_bytes', value} for sending binary data to endpoint?

Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Basically, the input format is very easy: the transform expects a PCollection of `tf.train.Example` objects, which are pretty well-known in tensorflow world. But I agree that the information about sending binary data is not visible enough. I'll try to improve it. 

> do you mean that we need to change l74 to something like:
feature={name+'_bytes', value} for sending binary data to endpoint?

Yes. It's a sign for the transform that data should be b64-encoded. In a request we'd have something like this:

`
"instances": [{"input_bytes": {"b64": "xxx"}}]
`

File: website/src/documentation/patterns/ai-platform.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  <strong>Adapt for:</strong>
+  <ul>
+    <li data-type="language-java">Java SDK</li>
+    <li data-type="language-py" class="active">Python SDK</li>
+  </ul>
+</nav>
+
+## Getting predictions
+
+This section shows how to use a cloud-hosted machine learning model to make predictions about new data using Google Cloud AI Platform Prediction within Beam's pipeline.
```
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: Is AI Platform the "cloud-hosted machine learning model"?

I think we can remove the last part ("within Beam's pipeline").

File: website/src/documentation/patterns/ai-platform.md (hunk 2 above)
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: To avoid the phrase "Beam's PTransform", I'd recommend something like "is a library with a Beam PTransform called `RunInterface`".

Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: "able to perform two types of inference"

I only see one type of inference discussed in this paragraph. If we're only addressing one type of inference here, then we don't need to mention that there's another type. A reader might expect there to be information on both types of inference.

Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: Add a comma before "which" in "which contains predictions"

Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: "One of them can use a service endpoint."

First state the inference you can do with the transform, then describe how to do it.

File: website/src/documentation/patterns/ai-platform.md (hunk 3 above)
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: If only TensorFlow models are supported, we should say "deploy a TensorFlow model to the cloud."

Also, does the cloud service to which you deploy the model have to be AI Platform? If so, we should say "deploy a TensorFlow model to AI Platform."

File: website/src/documentation/patterns/ai-platform.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library that provides `RunInference` Beam's PTransform. `RunInference` is a PTransform able to perform two types of inference. One of them can use a service endpoint. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for each element, sends a request to Google Cloud AI Platform Prediction service. The transform produces a PCollection of type `PredictLog` which contains predictions.
+
+Before getting started, deploy a machine learning model to the cloud. The cloud service manages the infrastructure needed to handle prediction requests in both efficient and scalable way. Only Tensorflow models are supported. For more information, see [Exporting a SavedModel for prediction](https://cloud.google.com/ai-platform/prediction/docs/exporting-savedmodel-for-prediction).
+
+Once a machine learning model is deployed, prepare a list of instances to get predictions for. 
+
+Here is an example of a pipeline that reads input instances from the file, converts JSON objects to `tf.train.Example` objects and sends data to the service. The content of a file can look like this:
```
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: "and sends data to the service"

"the service" is "AI Platform", right? We should state this explicitly.

File: website/src/documentation/patterns/ai-platform.md (hunk 6 above)
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: The first sentence of this paragraph is long and hard to follow. Consider splitting in up into two (or even three) shorter sentences.

File: website/src/documentation/patterns/overview.md [hunk 9]
```diff
@@ -38,6 +38,9 @@ Pipeline patterns demonstrate common Beam use cases. Pipeline patterns are based
 **Custom window patterns** - Patterns for windowing functions
 * [Using data to dynamically set session window gaps]({{ site.baseurl }}/documentation/patterns/custom-windows/#using-data-to-dynamically-set-session-window-gaps)
 
+**AI Platform integration patterns** - Patterns for Google AI Platform transforms
```
Comment (soyrice) [inline website/src/documentation/patterns/overview.md]: Use the full service name here: "Patterns for using the Google Cloud AI Platform Prediction"

Make sure the the service name is consistent throughout the pattern. I noticed that sometimes it's "Google Cloud AI Platform Prediction service", "Google Cloud AI Platform Prediction", etc.

I'd recommend using the full service name ("Google Cloud AI Platform Prediction") the first time its mentioned. Then, refer to the service with the shorthand "AI Platform Prediction". This way, you can use some shorthand but there's no ambiguity about the name of service.

File: website/src/documentation/patterns/ai-platform.md (hunk 7 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: No, AI Platform is a service that manages cloud-hosted machine learning models.
How about something like this?

> This section shows how to use Google Cloud AI Platform Prediction  to make predictions about new data from a cloud-hosted machine learning model.

I'll also put an url link to an overview of AI Platform Prediction.

File: website/src/documentation/patterns/ai-platform.md (hunk 2 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Fair enough, this is an another evidence that this mention causes confusion. I'll remote it.

File: website/src/documentation/patterns/ai-platform.md (hunk 3 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: Alright, but I'd rather keep this sentence: "only TensorFlow models are supported by the transform" anyway. AI Platform, unlike `RunInference`, doesn't only support TensorFlow models.  It's important to put an emphasis on that.

File: website/src/documentation/patterns/overview.md (hunk 9 above)
Comment (kamilwu) [reply website/src/documentation/patterns/overview.md]: Looks good. One minor comment: there are more Google Cloud AI Platform transforms. This pull request is just about Prediction, more patterns could be added later. So I'd stick with `Patterns for Google Cloud AI Platform transforms`

File: website/src/documentation/patterns/ai-platform.md (hunk 7 above)
Comment (soyrice) [reply website/src/documentation/patterns/ai-platform.md]: SG :)

File: website/src/documentation/patterns/ai-platform.md (hunk 3 above)
Comment (soyrice) [reply website/src/documentation/patterns/ai-platform.md]: SG - that definitely helps remove the ambiguity

Comment (soyrice) [reply website/src/documentation/patterns/ai-platform.md]: The new text says "`RunInference` is able to perform an inference that can use a service endpoint." Is the service endpoint for receiving data or sending requests? It might be helpful to clarify.

File: website/src/documentation/patterns/ai-platform.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    <li data-type="language-java">Java SDK</li>
+    <li data-type="language-py" class="active">Python SDK</li>
+  </ul>
+</nav>
+
+## Getting predictions
+
+This section shows how to use [Google Cloud AI Platform Prediction](https://cloud.google.com/ai-platform/prediction/docs/overview) to make predictions about new data from a cloud-hosted machine learning model.
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library with a Beam PTransform called `RunInference`. `RunInference` is able to perform an inference that can use a service endpoint. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for every bunch of elements, sends a request to AI Platform Prediction. The transform produces a PCollection of type `PredictLog`, which contains predictions.
```
Comment (soyrice) [inline website/src/documentation/patterns/ai-platform.md]: "for every bunch of elements, sends a request to AI Platform Prediction."

I assume this refers to batching data to AI Platform. Is there away to configure how many elements are in the batch? I think the previous version of this paragraph stated that one request is sent per element.

Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: >  Is there away to configure how many elements are in the batch?

No, in this case Beam handles the size of a batch on its own.

> I think the previous version of this paragraph stated that one request is sent per element.

That's right. After a discussion on batching it comes to my mind this is not truly correct.

File: website/src/documentation/patterns/ai-platform.md (hunk 3 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: In this case it's for receiving data. In other words, AI Platform Prediction exposes a service endpoint that can receive data. I'll clarify it.

File: website/src/documentation/patterns/ai-platform.md (hunk 10 above)
Comment (soyrice) [reply website/src/documentation/patterns/ai-platform.md]: Is there another doc (or some source code) we can link to that describes how Beam handles the batch size? It's probably not worth adding the extra details to this pattern but it may be helpful to provide a link for more information.

I also think we should use the term "batch" instead of "bunch" for additional clarity.

**Comment by soyrice** [comment]
Staged:
http://apache-beam-website-pull-requests.storage.googleapis.com/11075/documentation/patterns/ai-platform/index.html
http://apache-beam-website-pull-requests.storage.googleapis.com/11075/documentation/patterns/overview/index.html

**Comment by soyrice** [comment]
LGTM

File: website/src/documentation/patterns/ai-platform.md (hunk 10 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md]: I believe BatchElements (the transform responsible for batching here) doc is sufficient. Here's a [link](https://github.com/apache/beam/blob/42d79c29949725b3afabfb3754bfb394be594460/sdks/python/apache_beam/transforms/util.py#L540)

**Comment by kamilwu** [comment]
Thanks @soyrice for your in-depth review!

File: website/src/documentation/patterns/ai-platform.md (hunk 10 above)
Comment (soyrice) [reply website/src/documentation/patterns/ai-platform.md]: Nice. We should add the link to the page

**Comment by kamilwu** [comment]
> Maybe we need to wait until tfx_bsl is available as a public solution. @yifanmai @katsiapis -- please advise on when/how we should move forward.

Pinging again. @aaltay @yifanmai @katsiapis

**Comment by aaltay** [comment]
> > Maybe we need to wait until tfx_bsl is available as a public solution. @yifanmai @katsiapis -- please advise on when/how we should move forward.
> 
> Pinging again. @aaltay @yifanmai @katsiapis

Still an open question in active discussion within TFX group. I am guessing decision, and the changes on TFX will take some time.

And /cc @rose-rong-liu

**Comment by kamilwu** [comment]
Thanks, good to know there is an ongoing discussion. Please let me know if anything changes.

**Comment by kamilwu** [comment]
@yifanmai @katsiapis Is there any news about making tfx_bsl a public solution?

File: website/src/documentation/patterns/ai-platform.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  </ul>
+</nav>
+
+## Getting predictions
+
+This section shows how to use [Google Cloud AI Platform Prediction](https://cloud.google.com/ai-platform/prediction/docs/overview) to make predictions about new data from a cloud-hosted machine learning model.
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library with a Beam PTransform called `RunInference`. `RunInference` is able to perform an inference that can use an external service endpoint for receiving data. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for every batch of elements, sends a request to AI Platform Prediction. The size of a batch may vary. For more details on how Beam finds the best batch size, refer to a docstring for [BatchElements](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.util.html?highlight=batchelements#apache_beam.transforms.util.BatchElements).
+ 
+ The transform produces a PCollection of type `PredictLog`, which contains predictions.
```
Comment (rose-rong-liu) [inline website/src/documentation/patterns/ai-platform.md:39]: s/PredictLog/PredictionLog

File: website/src/documentation/patterns/ai-platform.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+from tfx_bsl.beam.run_inference import RunInference
+from tfx_bsl.proto import model_spec_pb2
+
+def convert_json_to_tf_example(json_obj):
+  dict_ = json.loads(json_obj)
+  for name, text in dict_.items():
+      value = tf.train.Feature(bytes_list=tf.train.BytesList(
+        value=[text.encode('utf-8')]))
+      feature = {name: value}
+      return tf.train.Example(features=tf.train.Features(feature=feature))
```
Comment (rose-rong-liu) [inline website/src/documentation/patterns/ai-platform.md:78]: Is there extra space?

File: website/src/documentation/patterns/ai-platform.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+with beam.Pipeline() as p:
+     _ = (p
+         | beam.io.ReadFromText('gs://my-bucket/samples.json')
+         | beam.Map(convert_json_to_tf_example)
+         | RunInference(
+             model_spec_pb2.InferenceEndpoint(
+                 model_endpoint_spec=model_spec_pb2.ModelEndpointSpec(
```
Comment (rose-rong-liu) [inline website/src/documentation/patterns/ai-platform.md:86]: ModelEndpointSpec will be changed to AIPlatformPredictionModelSpec in next release of tfx_bsl.

File: website/src/documentation/patterns/ai-platform.md (hunk 12 above)
Comment (kamilwu) [reply website/src/documentation/patterns/ai-platform.md:78]: No, I think it's fine. There's only one space, after the `return`.

**Comment by kamilwu** [comment]
Run RAT PreCommit

**Comment by rose-rong-liu** [comment]
Public tfx_bsl interface will be out in the next release (0.22) next week.

**Comment by kamilwu** [comment]
@rose-rong-liu Thanks! I'll merge the PR after the release.

**Comment by kamilwu** [comment]
tfx_bsl 0.22 is out.
@aaltay (and others as well) Do you see any further blockers or can I move forward with merge?

**Comment by aaltay** [comment]
> tfx_bsl 0.22 is out.
> @aaltay (and others as well) Do you see any further blockers or can I move forward with merge?

LGTM.

@rose-rong-liu - any concerns?

File: website/www/site/content/en/documentation/patterns/ai-platform.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library with a Beam PTransform called `RunInference`. `RunInference` is able to perform an inference that can use an external service endpoint for receiving data. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for every batch of elements, sends a request to AI Platform Prediction. The size of a batch may vary. For more details on how Beam finds the best batch size, refer to a docstring for [BatchElements](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.util.html?highlight=batchelements#apache_beam.transforms.util.BatchElements).
+ 
+ The transform produces a PCollection of type `PredictionLog`, which contains predictions.
+
+Before getting started, deploy a TensorFlow model to AI Platform Prediction. The cloud service manages the infrastructure needed to handle prediction requests in both efficient and scalable way. Do note that only TensorFlow models are supported by the transform. For more information, see [Exporting a SavedModel for prediction](https://cloud.google.com/ai-platform/prediction/docs/exporting-savedmodel-for-prediction).
```
Comment (rose-rong-liu) [inline website/www/site/content/en/documentation/patterns/ai-platform.md:32]: From the implementation side, it does not limit to tensorflow model, right? As cloud servers other model formats with the same predict API. Or is this for branding purpose?

**Comment by rose-rong-liu** [comment]
> @rose-rong-liu - any concerns?

Looks good to me. Thanks for the wait!

**Comment by aaltay** [comment]
@kamilwu -- @katsiapis had a few questions. Please do not merge yet, I will add his comments here.

File: website/www/site/content/en/documentation/patterns/ai-platform.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+This page describes common patterns in pipelines with Google Cloud AI Platform transforms.
+
+{{< language-switcher java py >}}
+
+## Getting predictions
+
+This section shows how to use [Google Cloud AI Platform Prediction](https://cloud.google.com/ai-platform/prediction/docs/overview) to make predictions about new data from a cloud-hosted machine learning model.
+ 
+[tfx_bsl](https://github.com/tensorflow/tfx-bsl) is a library with a Beam PTransform called `RunInference`. `RunInference` is able to perform an inference that can use an external service endpoint for receiving data. When using a service endpoint, the transform takes a PCollection of type `tf.train.Example` and, for every batch of elements, sends a request to AI Platform Prediction. The size of a batch may vary. For more details on how Beam finds the best batch size, refer to a docstring for [BatchElements](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.util.html?highlight=batchelements#apache_beam.transforms.util.BatchElements).
```
Comment (aaltay) [inline website/www/site/content/en/documentation/patterns/ai-platform.md:28]: `tf.train.Example` -> `tf.train.Example` or `tf.train.SequenceExample`

Comment (aaltay) [inline website/www/site/content/en/documentation/patterns/ai-platform.md:28]: The size of a batch may vary. -> The size of a batch is automatically computed.

File: website/www/site/content/en/documentation/patterns/ai-platform.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ 
+ The transform produces a PCollection of type `PredictionLog`, which contains predictions.
+
+Before getting started, deploy a TensorFlow model to AI Platform Prediction. The cloud service manages the infrastructure needed to handle prediction requests in both efficient and scalable way. Do note that only TensorFlow models are supported by the transform. For more information, see [Exporting a SavedModel for prediction](https://cloud.google.com/ai-platform/prediction/docs/exporting-savedmodel-for-prediction).
+
+Once a machine learning model is deployed, prepare a list of instances to get predictions for. To send binary data, make sure that the name of an input ends in `_bytes`. This will base64-encode data before sending a request.
```
Comment (aaltay) [inline website/www/site/content/en/documentation/patterns/ai-platform.md:34]: Is this still applicable? 

/cc @rose-rong-liu

File: website/www/site/content/en/documentation/patterns/ai-platform.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+import json
+
+import apache_beam as beam
+
+import tensorflow as tf
+from tfx_bsl.beam.run_inference import RunInference
+from tfx_bsl.proto import model_spec_pb2
+
+def convert_json_to_tf_example(json_obj):
+  dict_ = json.loads(json_obj)
```
Comment (aaltay) [inline website/www/site/content/en/documentation/patterns/ai-platform.md:62]: remove the underscore. dict_ -> dict

File: website/www/site/content/en/documentation/patterns/ai-platform.md (hunk 15 above)
Comment (kamilwu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md:28]: `tf.train.SequenceExample` is not supported for remote prediction: [link](https://github.com/tensorflow/tfx-bsl/blob/a49260a3935ef0f35456ec119a827b158eba3b2a/tfx_bsl/beam/run_inference.py#L406)

File: website/www/site/content/en/documentation/patterns/ai-platform.md (hunk 17 above)
Comment (kamilwu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md:62]: I changed to `samples`, `dict` is the name for dictionary build-in type.

File: website/www/site/content/en/documentation/patterns/ai-platform.md (hunk 14 above)
Comment (kamilwu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md:32]: IIRC yes, it does not limit to tensorflow model, as long as prediction input is valid (for instance, binary data is accepted only by tensorflow model)

File: website/www/site/content/en/documentation/patterns/ai-platform.md (hunk 16 above)
Comment (rose-rong-liu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md:34]: Yes, this is still applicable, and required by cloud inference [1].

[1] https://cloud.google.com/ai-platform/prediction/docs/online-predict#binary_data_in_prediction_input

File: website/www/site/content/en/documentation/patterns/ai-platform.md (hunk 15 above)
Comment (aaltay) [reply website/www/site/content/en/documentation/patterns/ai-platform.md]: @rose-rong-liu - any suggestions here?

Comment (rose-rong-liu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md]: We're actively looking for solutions for sequence example support, which we may leverage arrow on. Maybe add a note to mention sequence example will be supported shortly?

Comment (kamilwu) [reply website/www/site/content/en/documentation/patterns/ai-platform.md]: I think adding such a note makes sense. @aaltay Any thoughs?

**Comment by kamilwu** [comment]
I think every comment's been addressed, therefore I can merge the PR. Thanks everyone involved.
